autogpt/agent/base_ai_core.py

```diff
@@ -0,0 +1,132 @@
+            trio.MemorySendChannel, trio.MemoryReceiveChannel
+        ] = (None, None),
+        debug: bool = False,
+    ):
+        """
+        Initialize a class instance.
+
+        Args:
+            memory_provider (MemoryProvider): The memory provider.
+            llm_provider (LLMProvider): The LLM provider.
+            abilities (Abilities): The abilities that the AI can execute.
+            command_channel (typing.Tuple[trio.MemorySendChannel, trio.MemoryReceiveChannel], optional): The command channel.
+            event_channel (typing.Tuple[trio.MemorySendChannel, trio.MemoryReceiveChannel], optional): The event channel.
+            master_mind_id (str, optional): The master mind's id. Defaults to None.
+            master_mind_channel (typing.Tuple[trio.MemorySendChannel, trio.MemoryReceiveChannel], optional): The master mind's channel. Defaults to (None, None).
+
+        """
+        self.memory_provider = memory_provider
+        self.llm_provider = llm_provider
+        self.abilities = abilities
+        self.command_send_channel, self.command_receive_channel = command_channel
+        self.event_send_channel, self.event_receive_channel = event_channel
+
+        self.is_sub_mind = master_mind_id is not None
+        self.master_mind_id = master_mind_id
+        self.master_send_channel, self.master_receive_channel = master_mind_channel
+        self.debug = debug
+
+    async def send_event(self, event: Event) -> None:
+        """Send an event to the event channel.
+
+        Args:
+            event (Event): The event to be sent.
+        """
+        await self.event_send_channel.send(event)
+
+    async def receive_command(self) -> None:
+        """
+        Receive a command from the command channel and sends it for processing
+        """
+        command = await self.command_receive_channel.receive()
+        await self.process_commands(command)
+
+    async def receive_command_from_master(self) -> None:
+        """
+        Receive a command from the master mind's command channel and sends it for processing
+        """
+        command = await self.master_receive_channel.receive()
+        await self.process_commands(command)
+
+    async def send_report_to_master(self, command: Command) -> None:
+        """
+        Send a report to the master mind's event channel.
+        """
+        await self.master_send_channel.send(command)
+
+    @abc.abstractmethod
+    async def process_commands(self, command: Command) -> None:
+        """
+        Process a command.
+
+        Args:
+            command (Command): The command to be processed.
+        """
+        pass
+
+    async def run(self) -> None:
+        """The main loop of the AI core."""
+        keep_running = True
+        while keep_running:
+            await self.receive_command()
+            if self.is_sub_mind:
+                await self.receive_command_from_master()
+
+    # async def spawn_sub_mind(self, ai_core: BaseAICore) -> None:
```

Comment:
Curious to know why this is commented out.  I have been working on a crude ability to [clone the AutoGPT process ](https://github.com/KayLuke/Auto-GPT-Dolly-Plugin). This is way cleaner - it'd be a powerful feature...

Reply:
This PR is still in development. I commented this out whilst debugging an issue. It will be coming back :)